### ISSUE TITLE
SslStream AuthenticateAsClientAsync throws when running under netcoreapp2.0

### ISSUE THREAD

**Issue by darrensteadman** [dotnet/corefx#23332]
I have an app that I've been porting from net461 to netcoreapp2.0 over the last few days and I've come across a problem when using SslStream.

If I run the ported application as netcoreapp2.0 I get the following exception when running on Windows 7 x64

A call to SSPI failed, see inner exception.
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Net.Security.SslState.StartSendAuthResetSignal(ProtocolToken message, AsyncProtocolRequest asyncRequest, ExceptionDispatchInfo exception)
   at System.Net.Security.SslState.CheckCompletionBeforeNextReceive(ProtocolToken message, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartSendBlob(Byte[] incoming, Int32 count, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.ProcessReceivedBlob(Byte[] buffer, Int32 count, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartReadFrame(Byte[] buffer, Int32 readBytes, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartReceiveBlob(Byte[] buffer, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.CheckCompletionBeforeNextReceive(ProtocolToken message, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartSendBlob(Byte[] incoming, Int32 count, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.ForceAuthentication(Boolean receiveFirst, Byte[] buffer, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.ProcessAuthentication(LazyAsyncResult lazyResult)
   at System.Net.Security.SslStream.BeginAuthenticateAsClient(String targetHost, X509CertificateCollection clientCertificates, SslProtocols enabledSslProtocols, Boolean checkCertificateRevocation, AsyncCallback asyncCallback, Object asyncState)
   at System.Net.Security.SslStream.BeginAuthenticateAsClient(String targetHost, AsyncCallback asyncCallback, Object asyncState)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncImpl[TArg1](Func`4 beginMethod, Func`2 endFunction, Action`1 endAction, TArg1 arg1, Object state, TaskCreationOptions creationOptions)
   at System.Threading.Tasks.TaskFactory.FromAsync[TArg1](Func`4 beginMethod, Action`1 endMethod, TArg1 arg1, Object state, TaskCreationOptions creationOptions)

   at System.Threading.Tasks.TaskFactory.FromAsync[TArg1](Func`4 beginMethod, Action`1 endMethod, TArg1 arg1, Object state)
   at System.Net.Security.SslStream.AuthenticateAsClientAsync(String targetHost)

Inner exception details

System.Security.Authentication.AuthenticationException occurred
  HResult=0x80131501
  Message=A call to SSPI failed, see inner exception.
  Source=<Cannot evaluate the exception source>
  StackTrace:
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()

Inner Exception 1:
Win32Exception: The function requested is not supported

If I then run the same app on an linux-arm board I get the following exception

Unable to transfer data on the transport connection: Connection reset by peer.
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Net.FixedSizeReader.<ReadPacketAsync>d__1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Net.Security.SslState.InternalEndProcessAuthentication(LazyAsyncResult lazyResult)
   at System.Net.Security.SslState.EndProcessAuthentication(IAsyncResult result)
   at System.Net.Security.SslStream.EndAuthenticateAsClient(IAsyncResult asyncResult)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()

If I run the app on Windows 7 as net461 it all works as excepted.

On my Windows 7 machine I have OpenSSL installed and on path. It's version 1.0.2h
On the Linux board I have version 1.0.2d

In summary if I run the code as netcoreapp2.0 on linux-arm or Windows 7 I get an exception, if I run the same code but as net461 on Windows 7 everything works fine.

I'm using the official release of the tooling from dot.net

**Comment by darrensteadman** [comment]
I have investigated this further and believe I know what is happening.

If you call `AuthenticateAsClientAsync` with just the `targetHost` then internally when everything is setup `SecurityProtocol.SystemDefaultSecurityProtocols` is being used. This currently means that the SSL connection will be negotiated with `SslProtocols.None` set. This is basically why the above exceptions are being thrown because the thing at the other end of the connection doesn't support a protocol of 'None'. (Could probably do with a nicer exception that tells the developer that the other end doesn't accept the protocol)

According to the msdn documentation for `SslStream` [https://msdn.microsoft.com/en-us/library/system.net.security.sslstream.sslprotocol(v=vs.110).aspx](url) , `SslProtocol` is initialized to 'Default' which actually translates to `SslProtocol.Ssl3 | SslProtocol.Tls`, which is why everything works when running under net461. By default in netcoreapp2.0 `SslStream.SslProtocol` is being initialized to `SslProtocol.None`.

It would appear that SSL2 and SSL3 support are being dropped in the code from comments I saw in which case to keep compatibility with net461 I would have thought `SecurityProtocol.SystemDefaultSecurityProtocols` should be being set to `SslProtocol.Tls`.

If I call the version of `AuthenticateAsClientAsync` that allows me to specify the protocol and assign `SslProtocol.Tls | SslProtocol.Tls11 | SslProtocol.Tls12` then everything works fine in net461 and netappcore2.0

I'd see this as a break in default behavior from net461.

**Comment by davidsh** [comment]
>I'd see this as a break in default behavior from net461.

Yes, this is a change in behavior from .NET 4.6.1.  However, as of .NET 4.7, using SslProtocols.None (which implies using the most secure system defaults) is the new default behavior in .NET Framework.

This was a breaking change documented.

**Comment by infiniteloopltd** [comment]
If you are using SslStream, then you need to explicitly set the TLS version in the AuthenticateAsClient call, for example:

`ssl.AuthenticateAsClient(url, null, SslProtocols.Tls12, false);`
